sdk/core/azure-core/azure/core/tracing/decorator.py

```diff
@@ -48,30 +48,35 @@ def distributed_trace(func=None, name_of_span=None):
     @functools.wraps(func)
     def wrapper_use_tracer(*args, **kwargs):
         # type: (Any, Any) -> Any
+        merge_span = kwargs.pop('merge_span', False)
         passed_in_parent = kwargs.pop("parent_span", None)
-        orig_wrapped_span = tracing_context.current_span.get()
-        wrapper_class = settings.tracing_implementation()
-        original_span_instance = None
-        if wrapper_class is not None:
-            original_span_instance = wrapper_class.get_current_span()
+
+        span_impl_type = settings.tracing_implementation()
+        if span_impl_type is None:
+            return func(*args, **kwargs) # type: ignore
+
+        # Merge span is parameter is set, but only if no explicit parent are passed
+        if merge_span and not passed_in_parent:
+            return func(*args, **kwargs) # type: ignore
+
+        # Get original context
+        original_wrapped_span = tracing_context.current_span.get()  # type: AbstractSpan
+        original_span_instance = span_impl_type.get_current_span()
+
         parent_span = common.get_parent_span(passed_in_parent)
-        ans = None
-        if parent_span is not None and orig_wrapped_span is None:
+        common.set_span_contexts(parent_span)
+
+        name = name_of_span or common.get_function_and_class_name(func, *args)  # type: ignore
+        child = parent_span.span(name=name)
+        try:
+            with child:
+                common.set_span_contexts(child)
+                return func(*args, **kwargs) # type: ignore
+        finally:
             common.set_span_contexts(parent_span)
-            name = name_of_span or common.get_function_and_class_name(func, *args)  # type: ignore
-            child = parent_span.span(name=name)
-            child.start()
-            common.set_span_contexts(child)
-            try:
-                ans = func(*args, **kwargs) # type: ignore
-            finally:
-                child.finish()
-                common.set_span_contexts(parent_span)
-                if orig_wrapped_span is None and passed_in_parent is None and original_span_instance is None:
-                    parent_span.finish()
-                common.set_span_contexts(orig_wrapped_span, span_instance=original_span_instance)
-        else:
-            ans = func(*args, **kwargs) # type: ignore
-        return ans
+            # This test means "get_parent" created the span, so I need to finish it.
+            if original_wrapped_span is None and passed_in_parent is None and original_span_instance is None:
+                parent_span.finish()
+            common.set_span_contexts(original_wrapped_span, span_instance=original_span_instance)
 
     return wrapper_use_tracer
```

Comment:
The decorator functions have gotten pretty complicated, I think they would benefit from a comment like you added at the top of `get_parent` to give a high level outline

Reply:
I simplified the code a ton, should be easy to get now :)